### PR TITLE
Updated ensmallen

### PR DESCRIPTION
**Pull request by LucaCappelletti94**
Refactored library to support ensmallen version 0.4.0.

### CODE REVIEW

**Review by pnrobinson — CHANGES_REQUESTED** [review]
in ``node2vec.py`` i


1) It would be good to replace
```
 optimizer: Union[str, Optimizer] = "nadam",
```
with  
```
optimizer: Union[str, Optimizer] = tf.keras.optimizers.Nadam,
```
for clarity

2) 
  self._negatives_samples = negatives_samples
should be ``negative_samples`` (spelling error)

3)  window_size: 
-- explanation missingh from parameter doc

4) Documentation
  # If there is more than one value per single sample
        if self._get_true_input_length() > 1:
=> This is for CBOW, right? Should we add a comment about this here?

**Review by pnrobinson — CHANGES_REQUESTED** [review]
2) abstract_node2vec_sequence
spelling errors
 -- Wthever to shuffle the vectors. => "Whether..."
--   Weight on the probability of returning to node coming from
=>   Weight on the probability of returning to the same node the walk just came from
--  This object can be created using the method available from graph (? do you mean available from ensmallen?)
-- samples_number => should be ``sample_number``

File: embiggen/sequences/abstract_sequence.py
Comment (pnrobinson) [inline embiggen/sequences/abstract_sequence.py:40]: samples_number should be spelled ``sample_number``

File: embiggen/sequences/abstract_word2vec_sequence.py
Comment (pnrobinson) [inline embiggen/sequences/abstract_word2vec_sequence.py:56]: Wthever to shuffle the vectors
spelling error in doc string, should be Whether

File: embiggen/sequences/link_prediction_sequence.py
Comment (pnrobinson) [inline embiggen/sequences/link_prediction_sequence.py:87]: Spelling errors
-- (first line) Keras Sequence for running Neural Netwok on graph link prediction. => "Network"
--  Index corresponding to batch to be rendered. => "rendered" is not quite right here, "to be returned" is better

File: embiggen/sequences/node2vec_sequence.py
Comment (pnrobinson) [inline embiggen/sequences/node2vec_sequence.py:58]: spelling -- should be ``dense_node_mapping``

File: embiggen/visualizations/graph_visualizations.py
Comment (pnrobinson) [inline embiggen/visualizations/graph_visualizations.py:450]: Spelling
-- Return given arrats shuffled synchronously. (arrays)
--"""Reset the axes ticks and set the given title.""" => ``the axis ticks``
-- in _set_legend, "handles" should be "handlers" (param name and documentation), handle has a different meaning from handler
-- Documentation of function plot_node_degrees ("Plot common node types of provided graph.") is probably copy paste error?
-- line 426, egde_labels => edge_labels

**Review by pnrobinson — CHANGES_REQUESTED** [review]
The code looks great, it is really elegant. I have made a bunch of suggestions to correct spelling errors.

**Comment by LucaCappelletti94** [comment]
> in `node2vec.py` i
> 
> 1. It would be good to replace
> 
> ```
>  optimizer: Union[str, Optimizer] = "nadam",
> ```
> 
> with
> 
> ```
> optimizer: Union[str, Optimizer] = tf.keras.optimizers.Nadam,
> ```
> 
> for clarity

That would have a different meaning: `"nadam"` in Keras (passed as a string) is interpreted as 'create a new Nadam object with default parameters' and it's how you usually pass it in Keras models. The other version you propose `tf.keras.optimizers.Nadam` would be a class object. If you were to instead use `tf.keras.optimizers.Nadam()` to get an instance of the object, you would get considerable headaches because default parameters that are editable objects (like dictionaries or lists) behave like static objects that are shared among different instances of the function or method containing them.

> 
> 
> self._negatives_samples = negatives_samples
> should be `negative_samples` (spelling error)

Fixed!

> 
> 1. window_size:
>    -- explanation missingh from parameter doc

Added proper documentation for parameter.

> 2. Documentation
> 
> # If there is more than one value per single sample
> ```
>     if self._get_true_input_length() > 1:
> ```
> 
> => This is for CBOW, right? Should we add a comment about this here?

This is generally to choose to build the model as a CBOW model or a SkipGram model. I have extended the comments to try and make its logic more clear.

**Comment by LucaCappelletti94** [comment]
> 1. abstract_node2vec_sequence
>    spelling errors
>    -- Wthever to shuffle the vectors. => "Whether..."
Addressed typo.
>    --   Weight on the probability of returning to node coming from
>    =>   Weight on the probability of returning to the same node the walk just came from
Replaced.
>    --  This object can be created using the method available from graph (? do you mean available from ensmallen?)
Extended comment.
>    -- samples_number => should be `sample_number`
Fixed typo.

File: embiggen/sequences/abstract_sequence.py
Comment (LucaCappelletti94) [reply embiggen/sequences/abstract_sequence.py]: Fixed!

File: embiggen/sequences/abstract_word2vec_sequence.py
Comment (LucaCappelletti94) [reply embiggen/sequences/abstract_word2vec_sequence.py:56]: Fixed!

File: embiggen/sequences/link_prediction_sequence.py
Comment (LucaCappelletti94) [reply embiggen/sequences/link_prediction_sequence.py:87]: Fixed both.

File: embiggen/visualizations/graph_visualizations.py
Comment (LucaCappelletti94) [reply embiggen/visualizations/graph_visualizations.py:450]: 1. Fixed!
2. Within matplotlib the term `axes` is used to reference this type of object, is it a mistake in this context?
3. The term `handles` is the one used within matplotlib.
4. Fixed typo.
5. Fixed typo.

File: embiggen/sequences/node2vec_sequence.py
Comment (LucaCappelletti94) [reply embiggen/sequences/node2vec_sequence.py:58]: Fixed: since this is also a typo within ensmallen, it will break the current branch. It will be fixed in the next version of ensmallen that we should release this afternoon.

File: embiggen/visualizations/graph_visualizations.py
Comment (pnrobinson) [reply embiggen/visualizations/graph_visualizations.py:450]: axis=singular, axes plural, and so normally in English you would say axis ticks. matplotlib uses 'axes' (correctly) in other contexts, e.g., https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.set_xticks.html

**Comment by vidarmehr** [comment]
@LucaCappelletti94 Luca, shall we merge pull request?

**Comment by LucaCappelletti94** [comment]
Yes for me we can absolutely proceed, I was waiting for @pnrobinson seal of approval.

**Comment by vidarmehr** [comment]
@LucaCappelletti94 Thanks, Luca. I think you addressed all Peter's comments. Is there anything left that needs Peter's approval?

**Comment by LucaCappelletti94** [comment]
Just the pull requested is still missing his 'check', but if you think that we can proceed with everyone's approval I can merge and publish the updated version.

**Comment by pnrobinson** [comment]
I think we can move ahead, molte grazie!

**Pull request merged**
